### ISSUE TITLE
reactor-netty 0.7.7 leaks ClassLoader

### ISSUE THREAD

**Issue by ArloL** [mjiderhamn/classloader-leak-prevention#82]
Hi @mjiderhamn and thanks for all the work on this library!

I was just testing the newly released Spring Boot Admin 2.0 with our [tomcat-classloader-leak-test](https://github.com/evosec/tomcat-classloader-leak-test) and sadly it fails. Analyzing the heap dump gives me this picture:

![2018-05-29 14_19_58-eclipse memory analyzer](https://user-images.githubusercontent.com/195367/40658397-6b12a620-634b-11e8-93c3-5b6f54441417.png)

I set a breakpoint in [DefaultLoopResources$EventLoop#L238](https://github.com/reactor/reactor-netty/blob/master/src/main/java/reactor/ipc/netty/resources/DefaultLoopResources.java#L238) and saw a huge stacktrace ending in [HttpClient#138](https://github.com/reactor/reactor-netty/blob/master/src/main/java/reactor/ipc/netty/http/client/HttpClient.java#L138) where `HttpResources.get()` is called but never cleaned up after (as far as I can see) even though [HttpResources](https://github.com/reactor/reactor-netty/blob/master/src/main/java/reactor/ipc/netty/http/HttpResources.java) is `reactor.coreDisposable`.

Is there something that this library can do about this? I will try opening an issue with reactor-netty as well.

**Comment by ArloL** [comment]
A workaround is to call HttpResources.shutdown when the context is being shutdown:

````
@Bean
public DisposableBean shutdownHttpResources() {
  return HttpResources::shutdown;
}
````

**Comment by mjiderhamn** [comment]
Hi @ArloL  and thanks for the report. Sorry for not finding the time to look into this sooner.

I pushed some changes to a [reactor-netty](https://github.com/mjiderhamn/classloader-leak-prevention/tree/reactor-netty) branch, invoking `HttpResources.shutdown()` on shutdown. Care to give it a try?

Did you ever file a reactor-netty issue that I can reference?

Also, as I'm not familiar with Spring Boot Admin, would you care to give some context? This is not in itself a Spring Boot application, but a regular `.war` file deployed to an app server? Is reactor-netty contained within that `.war`?

**Comment by ArloL** [comment]
Hi @mjiderhamn, no worries. It was not too important for me either so it took me a while to investigate further.

I did test it but it sadly does not work. The fully qualified class name is `reactor.ipc.netty.http.HttpResources` instead of `reactor.netty.http.HttpResources`.

No, I did not. I want to check which project calls HttpResources.get() because IMO that code should cleanup the allocated resources.

[Spring Boot Admin](https://github.com/codecentric/spring-boot-admin/) is in itself a Spring Boot application. I have created an internal project that uses the code and produces a `.war` file. Yes, reactor-netty is in that `.war` since it is a transitive dependency of `spring-boot-admin-server` -> `spring-boot-starter-webflux` -> `spring-boot-starer-reactor-netty` -> `reactor-netty`.
